metafacture-json/src/main/java/org/metafacture/json/JsonValidator.java

```diff
@@ -0,0 +1,194 @@
+     */
+    public void setWriteInvalid(final String writeInvalid) {
+        this.writeInvalid = fileWriter(writeInvalid);
+    }
+
+    /**
+     * @param idKey The JSON key for the record ID value.
+     */
+    public void setIdKey(final String idKey) {
+        this.idKey = idKey;
+    }
+
+    @Override
+    public void process(final String json) {
+        try {
+            validate(json, new JSONObject(json) /* throws JSONException on syntax error */);
+        }
+        catch (final JSONException e) {
+            handleInvalid(json, null, e.getMessage());
+        }
+    }
+
+    private void validate(final String json, final JSONObject object) {
+        try {
+            schema.validate(object); // throws ValidationException if invalid
+            getReceiver().process(json);
+            ++success;
+            write(json, writeValid);
+        }
+        catch (final ValidationException e) {
+            handleInvalid(json, object, e.getAllMessages().toString());
+        }
+    }
+
+    @Override
+    protected void onCloseStream() {
+        close(writeInvalid);
+        close(writeValid);
+        LOG.debug("Success: {}, Fail: {}", success, fail);
+        super.onCloseStream();
+    }
+
+    private void initSchema(final String schemaUrl) {
+        if (schema != null) {
+            return;
+        }
+        SchemaLoaderBuilder schemaLoader = SchemaLoader.builder();
+        try {
+            final URL url = new URL(schemaUrl);
+            schemaLoader = schemaLoader.schemaJson(jsonFrom(url.openStream()))
+                    .resolutionScope(baseFor(url.toString()));
+        }
+        catch (final IOException e) {
+            LOG.info("Could not read as URL: {}, trying to load from class path", schemaUrl);
+            schemaLoader = schemaLoader.schemaClient(SchemaClient.classPathAwareClient())
+                    .schemaJson(jsonFrom(getClass().getResourceAsStream(schemaUrl)))
+                    .resolutionScope("classpath://" + baseFor(schemaUrl));
+        }
+        schema = schemaLoader.build().load().build();
+    }
+
+    private JSONObject jsonFrom(final InputStream inputStream) {
+        try {
+            return new JSONObject(new JSONTokener(inputStream));
+        }
+        catch (final JSONException e) {
+            throw new MetafactureException(e.getMessage(), e);
+        }
+    }
+
+    private String baseFor(final String path) {
+        return path.substring(0, path.lastIndexOf('/') + 1);
+    }
+
+    private FileWriter fileWriter(final String fileLocation) {
+        try {
+            return new FileWriter(fileLocation);
+        }
+        catch (final IOException e) {
+            throw new MetafactureException(e.getMessage(), e);
+        }
+    }
+
+    private void handleInvalid(final String json, final JSONObject object,
+            final String errorMessage) {
+        LOG.info("Invalid JSON: {} in {}", errorMessage, object != null ? object.opt(idKey) : json);
```

Comment:
Shouldn't this be logged at `WARN` level rather than `INFO`?

Reply:
For this one, I think I'd prefer to stay on INFO. It was a specific decision for the usage in OERSI, where the validator came from (see details on the [logging setup decisions](https://gitlab.com/oersi/oersi-etl/-/issues/53)), and I think it makes sense here too: the validator acts like a filter in workflows, passing on only the valid entries. For a use case where we create data and then validate it, invalid documents are unexpected, but this module could very well be used in a way where invalid documents are expected, e.g. processing only those input files that conform to a specific schema.

Reply:
Okay, thanks for the explanation.